rocAL/source/loaders/audio/audio_loader_sharded.cpp

```diff
@@ -0,0 +1,157 @@
+        return;
+    _shard_count = reader_cfg.get_shard_count();
+    // Create loader modules
+    for (size_t i = 0; i < _shard_count; i++) {
+        std::shared_ptr loader = std::make_shared<AudioLoader>(_dev_resources);
+        loader->set_prefetch_queue_depth(_prefetch_queue_depth);
+        _loaders.push_back(loader);
+    }
+    // Initialize loader modules
+    for (size_t idx = 0; idx < _shard_count; idx++) {
+        _loaders[idx]->set_output(_output_tensor);
+        _loaders[idx]->set_gpu_device_id(idx);
+        reader_cfg.set_shard_count(_shard_count);
+        reader_cfg.set_shard_id(idx);
+        _loaders[idx]->initialize(reader_cfg, decoder_cfg, mem_type, batch_size, keep_orig_size);
+    }
+    _initialized = true;
+}
+
+void AudioLoaderSharded::start_loading() {
+    for (unsigned i = 0; i < _loaders.size(); i++) {
+        _loaders[i]->start_loading();
+        //  Changing thread scheduling policy and it's priority does not help on latest Ubuntu builds
+        //  and needs tweaking the Linux security settings , can be turned on for experimentation
+#if 0
+        // Set thread scheduling policy
+        struct sched_param params;
+        params.sched_priority = sched_get_priority_max(SCHED_FIFO);
+        _loaders[i]->set_cpu_sched_policy(params);
+#endif
+        // Setting cpu affinity for threads works and can be activated below for experimentation
+#if 0
+        // Set thread affinity thread 0 to core 0 , 1 toc core 1 , ...
+        cpu_set_t cpuset;
+        CPU_ZERO(&cpuset);
+        CPU_SET(i, &cpuset);
+        _loaders[i]->set_cpu_affinity(cpuset);
+#endif
+    }
+}
+
+void AudioLoaderSharded::shut_down() {
+    for (unsigned i = 0; i < _loaders.size(); i++)
+        _loaders[i]->shut_down();
+}
+
+void AudioLoaderSharded::set_output(Tensor* output_tensor) {
+    _output_tensor = output_tensor;
+}
+
+size_t AudioLoaderSharded::remaining_count() {
+    int sum = 0;
+    for (auto& loader : _loaders)
+        sum += loader->remaining_count();
+    return sum;
+}
+
+void AudioLoaderSharded::reset() {
+    for (auto& loader : _loaders) {
+        loader->reset();
+    }
+}
+
+void AudioLoaderSharded::increment_loader_idx() {
+    _loader_idx = (_loader_idx + 1) % _shard_count;
+}
+
+Timing AudioLoaderSharded::timing() {
+    Timing t;
+    long long unsigned max_decode_time = 0;
+    long long unsigned max_read_time = 0;
+    long long unsigned swap_handle_time = 0;
+    // audio read and decode runs in parallel using multiple loaders, and the observable latency that the AudioLoaderSharded user
+    // is experiences on the load_next() call due to read and decode time is the maximum of all
+    for (auto& loader : _loaders) {
+        auto info = loader->timing();
+        max_read_time = (info.read_time > max_read_time) ? info.read_time : max_read_time;
+        max_decode_time = (info.decode_time > max_decode_time) ? info.decode_time : max_decode_time;
+        swap_handle_time += info.process_time;
```

Comment:
why is the process time baked into swap_handle_time. We should call it process_time instead?

Reply:
In rocAL the swap handle time is also added with the process time in addition to the graph process time in master graph
Renamed it to accumulated_process_time, since we are accumulating the time from all the loaders